### PR TITLE
[Controller] Abstract db client & move functions from Base to DB client

### PR DESCRIPTION
**Pull request by yonishelach**
1. Added an abstraction for the DB Client
2. Moved `to_orm_object`, `merge_into_orm_object` and `from_orm_object` to `SqlClient`.
3. Some lint fixes

### CODE REVIEW

File: controller/src/controller/db/__init__.py [hunk 1]
```diff
@@ -13,9 +13,10 @@
 # limitations under the License.
 
 from controller.config import config
-from controller.db.sqlclient import SqlClient
 
 client = None
 
 if config.db_type == "sql":
+    from controller.db.sql.sqlclient import SqlClient
```
Comment (guy1992l) [inline controller/src/controller/db/__init__.py:20]:
```suggestion
    from controller.db.sql import SqlClient
```

Comment (guy1992l) [inline controller/src/controller/db/__init__.py]:
```suggestion
    from controller.db.sql import SqlClient
```

File: controller/src/controller/db/__init__.py [hunk 2]
```diff
@@ -13,9 +13,10 @@
 # limitations under the License.
 
 from controller.config import config
-from controller.db.sqlclient import SqlClient
 
 client = None
 
 if config.db_type == "sql":
```
Comment (guy1992l) [inline controller/src/controller/db/__init__.py:26]: Add DataBaseType(Enum) with SQL = "sql". Put it here in the init file.

File: controller/src/controller/db/client.py [hunk 3]
```diff
@@ -0,0 +1,688 @@
...
+from abc import ABC, abstractmethod
+from typing import List, Optional, Type, Union
+
+import genai_factory.schemas as api_models
+
+
+class Client(ABC):
+    """
+    Base class for all DB clients
+    In order to implement a new DB client, inherit from this class and implement all the methods.
```
Comment (guy1992l) [inline controller/src/controller/db/client.py]: although I like doc strings, this line is kinda obvious :)

File: controller/src/controller/db/client.py [hunk 4]
```diff
@@ -0,0 +1,688 @@
...
+    """
+
+    @abstractmethod
+    def __init__(self):
+        pass
```
Comment (guy1992l) [inline controller/src/controller/db/client.py]: I don't see the point in forcing init method. You only force him to have an init without any arguments as inputs, which is not true as you can pass db configurations via the controller config and send them to the client in his init.

File: controller/src/controller/db/client.py [hunk 5]
```diff
@@ -0,0 +1,688 @@
...
+
+    @abstractmethod
+    def get_local_session(self):
+        """
+        Get a local session from the local session maker.
+        This is the session that is inserted into the API endpoints.
+        :return: The session.
+        """
```
Comment (guy1992l) [inline controller/src/controller/db/client.py]:
```suggestion
        """
        Get a local session from the local session maker.
        This is the session that is inserted into the API endpoints.
        
        :return: The session.
        """
```

File: controller/src/controller/db/client.py [hunk 6]
```diff
@@ -0,0 +1,688 @@
...
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def _from_db_object(obj, obj_class):
```
Comment (guy1992l) [inline controller/src/controller/db/client.py]: What about changing the name to
```suggestion
    def _to_schema_object(obj, schema: BaseOrWhateverWeHave):
```

Makes more sense as currently, I don't understand from what to what?

File: controller/src/controller/db/client.py [hunk 7]
```diff
@@ -0,0 +1,688 @@
...
+        """
+        Convert an ORM object to a schema object.
+        :param obj:       The DB object.
+        :param obj_class: The schema class.
+        :return: The schema object.
+        """
```
Comment (guy1992l) [inline controller/src/controller/db/client.py]: make sure to add new lines between doc sections:

```suggestion
        """
        Convert an ORM object to a schema object.
        
        :param obj:       The DB object.
        :param obj_class: The schema class.
        
        :return: The schema object.
        """
```

File: controller/src/controller/db/client.py [hunk 8]
```diff
@@ -0,0 +1,688 @@
...
+        Convert a schema object to a DB object.
+        :param obj:       The schema object.
+        :param obj_class: The DB class.
+        :return: The DB object.
+        """
+        pass
+
+    @abstractmethod
+    def create_tables(self, drop_old: bool = False, names: list = None):
```
Comment (guy1992l) [inline controller/src/controller/db/client.py]: Not all DBs are table based. Take example MongoDB which is document based. How about simply `create_db`?

**Review by guy1992l — CHANGES_REQUESTED** [review]
Very nice and sharp. Very few comments. In addition, I would rename `sqlclient.py` to `client.py` and sqldb.py to `db.py` as they are under a directory named `sql`.

File: controller/src/controller/db/client.py (hunk 4 above)
Comment (yonishelach) [reply controller/src/controller/db/client.py]: So the db configuration should be in a separate variable in the config?

Comment (guy1992l) [reply controller/src/controller/db/client.py]: well, yeah I mean every DB will probably have its own stuff. So we can put it under "db" field and everything there will be passed as `MyChosenClient(**config["db"])`

File: controller/src/controller/db/client.py (hunk 6 above)
Comment (yonishelach) [reply controller/src/controller/db/client.py]: this convert the db object into schema object

File: controller/src/controller/db/client.py [hunk 9]
```diff
@@ -19,31 +19,25 @@
...
         This is the session that is inserted into the API endpoints.
+
         :return: The session.
         """
         pass
 
     @staticmethod
     @abstractmethod
-    def _from_db_object(obj, obj_class):
+    def _to_schema_object(obj, obj_class: Type[api_models.Base]):
```
Comment (guy1992l) [inline controller/src/controller/db/client.py]:
```suggestion
    def _to_schema_object(obj, schema_class: Type[api_models.Base]) -> api_models.Base:
```

File: controller/src/controller/db/client.py [hunk 10]
```diff
@@ -53,8 +47,10 @@ def _from_db_object(obj, obj_class):
     def _merge_into_db_object(obj, obj_class):
         """
         Merge a schema object into an DB object.
```
Comment (guy1992l) [inline controller/src/controller/db/client.py]: can you explain what it means to merge? if I wanted to implement this abstract method, I wouldn't know what to do here. Abstract methods require more information so their implementer will know what is expected from them.

File: controller/src/controller/db/client.py [hunk 11]
```diff
@@ -64,18 +60,21 @@ def _merge_into_db_object(obj, obj_class):
     def _to_db_object(obj, obj_class):
```
Comment (guy1992l) [inline controller/src/controller/db/client.py]: add a generic type so when implementing the type hint of the db class will be sorted:

```python
from typing import TypeVar, Generic

DBObjectType = TypeVar("DBObjectType")

class DBClient(Generic[DBObjectType]):
    ...
```

Then you can hint:

```suggestion
    def _to_db_object(obj, db_class: Type[DBObjectType]) -> DBObjectType:
```

Notice when inheriting, you need to specify the type: `SQLDBClient(DBClient[SQLAlchemy.Something])`

File: controller/src/controller/db/client.py [hunk 12]
```diff
@@ -64,18 +60,21 @@ def _merge_into_db_object(obj, obj_class):
...
+
         :return: The DB object.
         """
         pass
 
     @abstractmethod
-    def create_tables(self, drop_old: bool = False, names: list = None):
+    def create_database(self, drop_old: bool = False, names: list = None):
         """
         Create the tables in the database.
```
Comment (guy1992l) [inline controller/src/controller/db/client.py]:
```suggestion
        Create a new database.
```

File: controller/src/controller/db/sql/sqlclient.py [hunk 13]
```diff
@@ -30,10 +30,10 @@ class SqlClient(Client):
     This is the SQL client that interact with the SQL database.
     """
 
-    def __init__(self, db_url: str, verbose: bool = False):
-        self.db_url = db_url
+    def __init__(self, url: str, verbose: bool = False):
```
Comment (guy1992l) [inline controller/src/controller/db/sql/sqlclient.py]: notice you give `"db_url"` in the config, and you pass `**config.db` to the init, so renaming to `url` will break it.

**Review by guy1992l — CHANGES_REQUESTED** [review]

File: controller/src/controller/db/client.py (hunk 10 above)
Comment (yonishelach) [reply controller/src/controller/db/client.py]: irrelevant, not abstract anymore

File: controller/src/controller/db/client.py (hunk 11 above)
Comment (yonishelach) [reply controller/src/controller/db/client.py]: irrelevant, not abstract anymore

**Review by guy1992l — APPROVED** [review]

**Pull request merged**